test/Microsoft.AspNetCore.Blazor.Build.Test/RenderingRazorIntegrationTest.cs

```diff
@@ -96,9 +96,10 @@ public void SupportsElementsAsStaticBlock()
         [Fact]
         public void CreatesSeparateMarkupFrameForEachTopLevelStaticElement()
         {
-            // The JavaScript-side rendering code relies on markup blocks representing
-            // exactly one self-contained element. This test is to show that we don't
-            // coalesce together markup blocks that happen to be next to each other.
+            // The JavaScript-side rendering code does not rely on this behavior. It supports
+            // inserting markup frames with arbitrary markup (e.g., multiple top-level elements
+            // or none). This test exists only as an observation of the current behavior rather
+            // than a promise that we never want to change it.
```

Comment:
Do you think it's desirable that we change this in the compiler? It would be straightforward to do.

Reply:
Perf-wise yes, the ideal would be for adjacent literal markup frames (and any adjacent literal text frames) to be coalesced. 

If you think this is a simple addition that would be neat, but I think it’s also a good improvement as-is.

Note that we shouldn’t treat text literals as markup unless they are coalesced into a markup frame, because text nodes on their own are much cheaper to render than markup frames.